File: S1MCPClient/src/tools/world.py
"""World tool — game state, locations, vehicles, logs, and saves."""

import json
from typing import Any, Dict
from mcp.types import Tool, TextContent

from ..tcp_client import TcpClient
from ..utils.logger import get_logger

logger = get_logger()

TOOL_NAME = "s1_world"

TOOL = Tool(
    name=TOOL_NAME,
    description=(
        "Query and manage world/game state. Choose an action:\n"
        "- get_state: Current scene, game time, network mode, and loaded mods\n"
        "- list_locations: List all properties/buildings in the game\n"
        "- get_location: Get a specific property (requires: property_id or property_name)\n"
        "- list_vehicles: List all vehicles in the world\n"
        "- get_vehicle: Get vehicle details (requires: vehicle_id)\n"
        "- capture_logs: Read game logs from MelonLoader (optional: last_n_lines, first_n_lines, "
        "keyword, from_timestamp, to_timestamp, include_pattern, exclude_pattern)\n"
        "- list_saves: List all save game slots\n"
        "- load_save: Load a save by slot index (requires: slot_index, 0-based)"
    ),
    inputSchema={
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": [
                    "get_state", "list_locations", "get_location",
                    "list_vehicles", "get_vehicle",
                    "capture_logs", "list_saves", "load_save"
                ],
                "description": "Operation to perform"
            },
            "property_id": {
                "type": "string",
                "description": "Property unique identifier (get_location)"
            },
            "property_name": {
                "type": "string",
                "description": "Property name as alternative to property_id (get_location)"
            },
            "vehicle_id": {
                "type": "string",
                "description": "Vehicle unique identifier (get_vehicle)"
            },
            "slot_index": {
                "type": "integer",
                "minimum": 0,
                "description": "0-based save slot index (load_save)"
            },
            "last_n_lines": {
                "type": "integer",
                "minimum": 1,
                "description": "Return last N log lines (capture_logs)"
            },
            "first_n_lines": {
                "type": "integer",
                "minimum": 1,
                "description": "Return first N log lines (capture_logs)"
            },
            "keyword": {
                "type": "string",
                "description": "Filter logs by keyword (capture_logs)"
            },
            "from_timestamp": {
                "type": "string",
                "description": "Filter logs from timestamp HH:mm:ss (capture_logs)"
            },
            "to_timestamp": {
                "type": "string",
                "description": "Filter logs to timestamp HH:mm:ss (capture_logs)"
            },
            "include_pattern": {
                "type": "string",
                "description": "Regex pattern — only matching lines (capture_logs)"
            },
            "exclude_pattern": {
                "type": "string",
                "description": "Regex pattern — exclude matching lines (capture_logs)"
            }
        },
        "required": ["action"]
    }
)


async def handle(arguments: Dict[str, Any], tcp_client: TcpClient) -> list[TextContent]:
    action = arguments.get("action")

    try:
        if action == "get_state":
            resp = await tcp_client.async_call("get_game_state", {})

        elif action == "list_locations":
            resp = await tcp_client.async_call("list_properties", {})

        elif action == "get_location":
            pid = arguments.get("property_id")
            pname = arguments.get("property_name")
            if not pid and not pname:
                return [TextContent(type="text", text="Error: property_id or property_name is required")]
            params = {}
            if pid:
                params["property_id"] = pid
            if pname:
                params["property_name"] = pname
            resp = await tcp_client.async_call("get_property", params)

        elif action == "list_vehicles":
            resp = await tcp_client.async_call("list_vehicles", {})

        elif action == "get_vehicle":
            vid = arguments.get("vehicle_id")
            if not vid:
                return [TextContent(type="text", text="Error: vehicle_id is required")]
            resp = await tcp_client.async_call("get_vehicle", {"vehicle_id": vid})

        elif action == "capture_logs":
            params = {k: arguments[k] for k in (
                "last_n_lines", "first_n_lines", "keyword",
                "from_timestamp", "to_timestamp", "include_pattern", "exclude_pattern"
            ) if k in arguments}
            resp = await tcp_client.async_call("capture_logs", params)
            if resp.error:
                return [TextContent(type="text", text=f"Error: {resp.error.message} (code: {resp.error.code})")]
            return _format_logs(resp.result)

        elif action == "list_saves":
            resp = await tcp_client.async_call("list_saves", {})
            if resp.error:
                return [TextContent(type="text", text=f"Error: {resp.error.message} (code: {resp.error.code})")]
            result_text = json.dumps(resp.result, indent=2)
            if isinstance(resp.result, dict) and "saves" in resp.result:
                count = resp.result.get("count", 0)
                return [TextContent(type="text", text=f"Found {count} save(s):\n\n{result_text}")]
            return [TextContent(type="text", text=result_text)]

        elif action == "load_save":
            slot = arguments.get("slot_index")
            if slot is None:
                return [TextContent(type="text", text="Error: slot_index is required")]
            resp = await tcp_client.async_call("load_save", {"slot_index": slot})
            if resp.error:
                return [TextContent(type="text", text=f"Error: {resp.error.message} (code: {resp.error.code})")]
            if isinstance(resp.result, dict):
                msg = resp.result.get("message", "")
                note = "\n  Note: Returned to menu first." if resp.result.get("returned_to_menu") else ""
                return [TextContent(type="text", text=f"✓ {msg}{note}")]
            return [TextContent(type="text", text=json.dumps(resp.result, indent=2))]

        else:
            return [TextContent(type="text", text=f"Error: Unknown action '{action}'")]

        if resp.error:
            return [TextContent(type="text", text=f"Error: {resp.error.message} (code: {resp.error.code})")]
        return [TextContent(type="text", text=json.dumps(resp.result, indent=2))]

    except Exception as e:
        logger.error(f"Error in s1_world/{action}: {e}")
        return [TextContent(type="text", text=f"Error: {e}")]


def _format_logs(result) -> list[TextContent]:
    if not isinstance(result, dict):
        return [TextContent(type="text", text=str(result))]

    lines = result.get("lines", [])
    total = result.get("total_lines_in_file", 0)
    filtered = result.get("filtered_count", 0)
    filters = result.get("filters_applied", [])
    warning = result.get("warning")

    out = []
    if warning:
        out.append(f"Warning: {warning}\n")
    out.append(f"Log Summary: {len(lines)} lines returned ({filtered} after filters, {total} total)")
    if filters:
        out.append("Filters: " + ", ".join(filters))
    if lines:
        out.append("\n--- Log Lines ---")
        for lo in lines:
            ln = lo.get("line_number", "?")
            ts = lo.get("timestamp", "")
            content = lo.get("content", "")
            out.append(f"[{ln}]{f' [{ts}]' if ts else ''} {content}")
    else:
        out.append("(No lines matched)")
    return [TextContent(type="text", text="\n".join(out))]


TOOL_HANDLERS = {TOOL_NAME: handle}


def get_tools(_tcp_client: TcpClient) -> list[Tool]:
    return [TOOL]

File: S1MCPClientRust/src/tools/world.rs
use rmcp::model::{CallToolResult, JsonObject};
use serde_json::{json, Value};
use tracing::error;

use crate::mcp::tools::WorldArgs;
use crate::server_state::ServerState;

use super::common::{call_and_format_result, python_like_text, text_result};

const TOOL_NAME: &str = "s1_world";
const INTERNAL_ERROR_MESSAGE: &str = "An internal error occurred while processing your request.";

pub async fn handle_world(arguments: Option<JsonObject>, state: &ServerState) -> CallToolResult {
    let raw = Value::Object(arguments.unwrap_or_default());
    let args: WorldArgs = match serde_json::from_value(raw) {
        Ok(a) => a,
        Err(e) => return text_result(format!("Error: invalid arguments: {e}")),
    };

    match args.action.as_str() {
        "get_state" => {
            call_and_format_result(state, "get_game_state", Some(json!({})), TOOL_NAME, "get_state").await
        }
        "list_locations" => {
            call_and_format_result(
                state,
                "list_properties",
                Some(json!({})),
                TOOL_NAME,
                "list_locations",
            )
            .await
        }
        "get_location" => {
            if args.property_id.is_none() && args.property_name.is_none() {
                return text_result("Error: property_id or property_name is required");
            }
            let mut params = serde_json::Map::new();
            if let Some(id) = args.property_id {
                params.insert("property_id".to_string(), json!(id));
            }
            if let Some(name) = args.property_name {
                params.insert("property_name".to_string(), json!(name));
            }
            call_and_format_result(
                state,
                "get_property",
                Some(Value::Object(params)),
                TOOL_NAME,
                "get_location",
            )
            .await
        }
        "list_vehicles" => {
            call_and_format_result(
                state,
                "list_vehicles",
                Some(json!({})),
                TOOL_NAME,
                "list_vehicles",
            )
            .await
        }
        "get_vehicle" => {
            let Some(vehicle_id) = args.vehicle_id else {
                return text_result("Error: vehicle_id is required");
            };
            call_and_format_result(
                state,
                "get_vehicle",
                Some(json!({ "vehicle_id": vehicle_id })),
                TOOL_NAME,
                "get_vehicle",
            )
            .await
        }
        "capture_logs" => {
            let mut params = serde_json::Map::new();
            if let Some(v) = args.last_n_lines {
                params.insert("last_n_lines".to_string(), json!(v));
            }
            if let Some(v) = args.first_n_lines {
                params.insert("first_n_lines".to_string(), json!(v));
            }
            if let Some(v) = args.keyword {
                params.insert("keyword".to_string(), json!(v));
            }
            if let Some(v) = args.from_timestamp {
                params.insert("from_timestamp".to_string(), json!(v));
            }
            if let Some(v) = args.to_timestamp {
                params.insert("to_timestamp".to_string(), json!(v));
            }
            if let Some(v) = args.include_pattern {
                params.insert("include_pattern".to_string(), json!(v));
            }
            if let Some(v) = args.exclude_pattern {
                params.insert("exclude_pattern".to_string(), json!(v));
            }

            let response =
                match state.tcp_client().async_call("capture_logs", Some(Value::Object(params))).await {
                    Ok(r) => r,
                    Err(e) => {
                        error!(tool = TOOL_NAME, action = "capture_logs", error = %e, "Tool request failed");
                        return text_result(INTERNAL_ERROR_MESSAGE);
                    }
                };

            if let Some(err) = response.error {
                return text_result(format!("Error: {} (code: {})", err.message, err.code));
            }
            format_logs(response.result.unwrap_or(Value::Null))
        }
        "list_saves" => {
            let response =
                match state.tcp_client().async_call("list_saves", Some(json!({}))).await {
                    Ok(r) => r,
                    Err(e) => {
                        error!(tool = TOOL_NAME, action = "list_saves", error = %e, "Tool request failed");
                        return text_result(INTERNAL_ERROR_MESSAGE);
                    }
                };

            if let Some(err) = response.error {
                return text_result(format!("Error: {} (code: {})", err.message, err.code));
            }

            let result_value = response.result.unwrap_or(Value::Null);
            let result_text = match serde_json::to_string_pretty(&result_value) {
                Ok(t) => t,
                Err(e) => {
                    error!(tool = TOOL_NAME, action = "list_saves", error = %e, "Failed to serialize tool response");
                    return text_result(INTERNAL_ERROR_MESSAGE);
                }
            };

            if let Some(obj) = result_value.as_object() {
                if obj.contains_key("saves") {
                    let count = obj
                        .get("count")
                        .map(python_like_text)
                        .unwrap_or_else(|| "0".to_string());
                    return text_result(format!("Found {} save(s):\n\n{}", count, result_text));
                }
            }

            text_result(result_text)
        }
        "load_save" => {
            let Some(slot_index) = args.slot_index else {
                return text_result("Error: slot_index is required");
            };

            let response = match state
                .tcp_client()
                .async_call("load_save", Some(json!({ "slot_index": slot_index })))
                .await
            {
                Ok(r) => r,
                Err(e) => {
                    error!(tool = TOOL_NAME, action = "load_save", error = %e, "Tool request failed");
                    return text_result(INTERNAL_ERROR_MESSAGE);
                }
            };

            if let Some(err) = response.error {
                return text_result(format!("Error: {} (code: {})", err.message, err.code));
            }

            let result_value = response.result.unwrap_or(Value::Null);
            if let Some(obj) = result_value.as_object() {
                let message = obj
                    .get("message")
                    .and_then(Value::as_str)
                    .unwrap_or_default();
                let note = if obj
                    .get("returned_to_menu")
                    .and_then(Value::as_bool)
                    .unwrap_or(false)
                {
                    "\n  Note: Returned to menu first."
                } else {
                    ""
                };
                return text_result(format!("✓ {}{}", message, note));
            }

            match serde_json::to_string_pretty(&result_value) {
                Ok(t) => text_result(t),
                Err(e) => {
                    error!(tool = TOOL_NAME, action = "load_save", error = %e, "Failed to serialize tool response");
                    text_result(INTERNAL_ERROR_MESSAGE)
                }
            }
        }
        other => text_result(format!("Error: Unknown action '{other}'")),
    }
}

fn format_logs(result: Value) -> CallToolResult {
    let Some(obj) = result.as_object() else {
        return text_result(python_like_text(&result));
    };

    let lines = obj
        .get("lines")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let total_lines = obj
        .get("total_lines_in_file")
        .and_then(Value::as_u64)
        .unwrap_or(0);
    let filtered_count = obj
        .get("filtered_count")
        .and_then(Value::as_u64)
        .unwrap_or(0);
    let filters = obj
        .get("filters_applied")
        .and_then(Value::as_array)
        .map(|items| items.iter().map(python_like_text).collect::<Vec<String>>())
        .unwrap_or_default();
    let warning = obj.get("warning").and_then(Value::as_str);

    let mut output = Vec::new();
    if let Some(w) = warning {
        output.push(format!("Warning: {}\n", w));
    }

    output.push(format!(
        "Log Summary: {} lines returned ({} after filters, {} total)",
        lines.len(),
        filtered_count,
        total_lines
    ));

    if !filters.is_empty() {
        output.push(format!("Filters: {}", filters.join(", ")));
    }

    if lines.is_empty() {
        output.push("(No lines matched)".to_string());
        return text_result(output.join("\n"));
    }

    output.push("\n--- Log Lines ---".to_string());

    for line in lines {
        if let Some(line_obj) = line.as_object() {
            let line_number = line_obj
                .get("line_number")
                .map(python_like_text)
                .unwrap_or_else(|| "?".to_string());
            let timestamp = line_obj
                .get("timestamp")
                .and_then(Value::as_str)
                .unwrap_or_default();
            let content = line_obj
                .get("content")
                .map(python_like_text)
                .unwrap_or_default();

            if timestamp.is_empty() {
                output.push(format!("[{}] {}", line_number, content));
            } else {
                output.push(format!("[{}] [{}] {}", line_number, timestamp, content));
            }
        } else {
            output.push(format!("[?] {}", python_like_text(&line)));
        }
    }

    text_result(output.join("\n"))
}
